src/base/format_impl.h

```diff
@@ -0,0 +1,1016 @@
+			// The easy case: Just start writing.
+			size_t written = 0;
+			for (const char* a = arg; *a && written < precision_; ++a, ++out, ++written) {
+				*out = *a;
+			}
+			if (written < min_width_) {
+				written = min_width_ - written;
+				for (; written; ++out, --written) {
+					*out = ' ';
+				}
+			}
+			return out;
+		}
+
+		// The more complex case: We want a right-aligned string with a given minimum width.
+		const size_t arg_len = std::min<size_t>(precision_, strlen(arg));
+		if (arg_len < min_width_) {
+			for (size_t i = min_width_ - arg_len; i; ++out, --i) {
+				*out = ' ';
+			}
+		}
+		for (size_t l = arg_len; l; --l, ++out, ++arg) {
+			*out = *arg;
+		}
+		return out;
+	}
+
+	static const StringNode node_;
+};
+
+struct BooleanNode : StringNode {
+	BooleanNode(const uint8_t f, const size_t w, const int32_t p) : StringNode(f, w, p) {
+	}
+
+	char* append(char* out, const ArgType t, Argument arg, const bool localize) const override {
+		bool val;
+		switch (t) {
+		case ArgType::kBoolean:
+			val = arg.boolean_val;
+			break;
+		case ArgType::kSigned:
+			val = arg.signed_val != 0;
+			break;
+		case ArgType::kUnsigned:
+		case ArgType::kPointer:
+			val = arg.unsigned_val != 0;
+			break;
+		case ArgType::kNullptr:
+			val = false;
+			break;
+		case ArgType::kChar:
+			val = arg.char_val != 0;
+			break;
+		default:
+			throw wexception("Wrong argument type: excepted bool, found %s", to_string(t).c_str());
+		}
+		arg.string_val = localize ? val ? _("true") : _("false") : val ? "true" : "false";
+		return StringNode::append(out, ArgType::kString, arg, localize);
+	}
+
+	static const BooleanNode node_;
+};
+
+template <typename Number> struct NumberNodeT : FormatNode {
+	NumberNodeT(const uint8_t f, const size_t w, const bool hex, const bool uc)
+	   : FormatNode(f, w, 0), base_(hex ? 16 : 10), hexadecimal_(hex), uppercase_(uc) {
+		if ((flags_ & kLeftAlign) && (flags_ & kPadWith0)) {
+			throw wexception("'-' and '0' can not be combined");
+		}
+	}
+
+	char* append(char* out, const ArgType t, Argument arg_u, const bool localize) const override {
+		Number arg;
+		switch (t) {
+		case ArgType::kSigned:
+			arg = arg_u.signed_val;
+			break;
+		case ArgType::kUnsigned:
+		case ArgType::kPointer:
+			arg = arg_u.unsigned_val;
+			break;
+		case ArgType::kFloat:
+			arg = arg_u.float_val;
+			break;
+		case ArgType::kNullptr:
+			arg_u.string_val = "nullptr";
+			return StringNode::node_.append(out, AbstractNode::ArgType::kString, arg_u, localize);
+		default:
+			throw wexception("Wrong argument type: excepted %s, found %s",
+			                 (std::is_signed<Number>::value ? "int" : "unsigned"),
+			                 to_string(t).c_str());
+		}
+
+		if (min_width_ == 0 || flags_ & kLeftAlign) {
+			// The easy case: Just start writing.
+			size_t written = 0;
+			if (arg < 0) {
+				*out = '-';
+				++out;
+				++written;
+			} else if (flags_ & kNumberSign) {
+				*out = '+';
+				++out;
+				++written;
+			}
```

Comment:
Shouldn't `0` be signless, but padded with space with the `+` flag, at least for integers?
I guess it's the same in the C standard, but why not improve it if you reimplement it anyway.

And as you're doing localised decimal separator for floats:
I think the `+` flag is most often used to keep alignment of digits, but plain ASCII `-` (and space) is narrower than digits and `+` (which are generally kept the same width) in variable width fonts, so the unicode minus sign (and I guess there must be a unicode as-wide-as-digits-space character) could be considered here.

Reply:
The non-localized bformat is often used for internal purposes, where we need to be able to assume that we use only ASCII characters. Also 0 is technically a positive number in C. 

For localized string though, I changed it to use the Unicode Minus sign `−` and for 0 the Unicode plusminus sign `±` how about that?

Reply:
Thank you.

For `0`, I found what I meant: U+2007 figure space. If that's supported by the fonts used in Widelands, then it would be much more appropriate than `±`.